### ISSUE TITLE
Add support for stateful loss functions

### ISSUE THREAD

**Issue by MaxSavenkov** [tiny-dnn/tiny-dnn#861]
I'm playing around with reinforcement learning, and in this case, loss value should only be calculated for one output, and be zero for the rest. It's not much of a problem in most cases, as desired_outputs parameter in fit<> call can be manipulated to make all irrelevant outputs zero, however, this is not the case when using cross_entropy_multiclass for training policy estimator in Actor-Critic method.

The problem is, output of softmax layer can reach 0, or be very close to it. But df function cross_entropy_multiclass is not ready to handle this case, and proceeds to divide by zero.

I think it would be useful to be able to pass loss function object instead of loss function class to fit<> in this case, so I could specify index for which training is performed, and only calculate loss for it, and leave the rest of the loss vector zero. This would prevent both unnecessary work and possible errors. Currently, it is not possible with tiny_dnn.

**Comment by MaxSavenkov** [comment]
Even more interesting is the case where you have to have different loss function for different outputs. Once again, this is what happens in Actor-Critic, when it is implemented as described here (https://medium.com/emergent-future/simple-reinforcement-learning-with-tensorflow-part-8-asynchronous-actor-critic-agents-a3c-c88f72a5e9f2), as a network with two "heads" (for policy and value function approximation). To be trained, one needs something like cross_entropy_multiclass, and the other MSE. Tensorflow API sidesteps the issue by allowing you to train the network with pre-computed gradients, but tiny-dnn does not allow this approach, as far as I can see. As an alternative to pre-computing gradients, gradient function in tiny_dnn might be modified to receive channel index to discriminate between types of loss to be computed.

**Comment by bhack** [comment]
Join the chat to discuss this topic a little bit: https://gitter.im/tiny-dnn/developers

**Comment by zulman** [comment]
+1
